simple_producer.go

```diff
@@ -26,47 +23,36 @@ func NewSimpleProducer(client *Client, config *ProducerConfig) (*SimpleProducer,
 		return nil, err
 	}
 
-	sp := &SimpleProducer{
-		producer:        prod,
-		newExpectations: make(chan *spExpect), // this must be unbuffered
-	}
+	sp := &SimpleProducer{producer: prod}
 
-	go withRecover(sp.matchResponses)
+	sp.wg.Add(2)
+	go withRecover(sp.handleSuccesses)
+	go withRecover(sp.handleErrors)
 
 	return sp, nil
 }
 
 // SendMessage produces a message to the given topic with the given key and value. To send strings as either key or value, see the StringEncoder type.
 func (sp *SimpleProducer) SendMessage(topic string, key, value Encoder) error {
-	msg := &MessageToSend{Topic: topic, Key: key, Value: value}
-	expectation := &spExpect{msg: msg, result: make(chan error)}
-	sp.newExpectations <- expectation
+	expectation := make(chan error)
+	msg := &MessageToSend{Topic: topic, Key: key, Value: value, Metadata: expectation}
 	sp.producer.Input() <- msg
-
-	return <-expectation.result
+	return <-expectation
 }
 
-func (sp *SimpleProducer) matchResponses() {
-	newExpectations := sp.newExpectations
-	unmatched := make(map[*MessageToSend]chan error)
-	unmatched[nil] = nil // prevent it from emptying entirely
+func (sp *SimpleProducer) handleSuccesses() {
+	defer sp.wg.Done()
+	for msg := range sp.producer.Successes() {
+		expectation := msg.Metadata.(chan error)
+		close(expectation)
```

Comment:
food for thought: should this be `expectation <- nil` instead? That would be a little clearer and more consistent with the `handleErrors` case...?